nowcasting_dataloader/utils/position_encoding.py

```diff
@@ -0,0 +1,291 @@
+            shape, as it has the most elements and the input (1, 6) can just subselect the position encoding
+        geospatial_coordinates: The latitude/longitude of the inputs for shape, in OSGB coordinates
+        datetimes: time of day and date for each of the timesteps in the shape
+        method: Method of the encoding, either 'fourier' for Fourier Features
+        geospatial_bounds: The bounds of the geospatial area covered, in a dict with the keys 'x_min', 'y_min', 'x_max', 'y_max'
+        kwargs: Passed to fourier_encode
+
+    Returns:
+        The position encodings for all items in the batch
+    """
+    assert method in [
+        "fourier",
+    ], AssertionError(f"method must be one of 'fourier', not '{method}'")
+
+    position_encoding = encode_absolute_position(
+        shape, geospatial_coordinates, geospatial_bounds, datetimes, **kwargs
+    )
+    return position_encoding
+
+
+def encode_absolute_position(
+    shape: List[int],
+    geospatial_coordinates: List[np.ndarray],
+    geospatial_bounds: Dict[str, float],
+    datetimes: List[datetime.datetime],
+    **kwargs,
+) -> torch.Tensor:
+    """
+    Encodes the absolute position of the pixels/voxels in time and space
+
+    This should be done per-modality and can be thought of as the relative position of the input modalities across a
+    given year and the area of the Earth covered by all the examples.
+
+    Args:
+        shape: Shape to encode positions for
+        geospatial_coordinates: The geospatial coordinates, in OSGB format
+        datetimes: Time of day and date as a list of datetimes, one for each timestep
+        geospatial_bounds: The geospatial bounds of the area where the examples come from, e.g. the coordinates of the area covered by the SEVIRI RSS image
+        **kwargs:
+
+    Returns:
+        The absolute position encoding for the given shape
+    """
+    datetime_features = create_datetime_features(datetimes)
+
+    # Fourier Features of absolute position
+    encoded_latlon = normalize_geospatial_coordinates(
+        geospatial_coordinates, geospatial_bounds, **kwargs
+    )
+
+    # Combine time and space features
+    to_concat = [einops.repeat(encoded_latlon, "b h w c -> b c t h w", t=shape[1])]
+    for date_feature in datetime_features:
+        to_concat.append(
+            einops.repeat(date_feature, "b t -> b c t h w", h=shape[-2], w=shape[-1], c=1)
+        )
+
+    # Now combined into one large encoding
+    absolute_position_encoding = torch.cat(to_concat, dim=1)
+
+    return absolute_position_encoding
+
+
+def normalize_geospatial_coordinates(
+    geospatial_coordinates: List[np.ndarray], geospatial_bounds: Dict[str, float], **kwargs
+) -> torch.Tensor:
+    """
+    Normalize the geospatial coordinates by the max extant to keep everything between -1 and 1, in sin and cos
+
+    This should work on a batch level, as the geospatial bounds should be the same for every example in the batch
```

Comment:
Ooooh... interesting....

I wonder if, actually, the position encoding (for both time and space) should be normalised so, for a given "real" location on Earth, the ML model always sees exactly the same position encoding??  My thinking is that the model might find it useful to learn that, say, clouds in Cornwall behave slightly differently to clouds in Scotland?

Having said all that, I think the code already works in this way?!  So maybe all that needs to be done here is maybe slightly clarify the docstring?

Reply:
That all sounds good to me!  I think the code is great as it is.  Maybe all that's required is just tweaking the docstring (I read the docstring as meaning that the normalisation would be done on a batch-by-batch basis) if that's OK?